### PR TITLE
[WIP] Add 781

### PR DESCRIPTION
**Pull request by PGijsbers**
Add a pre-commit hook with flake8, mypy and black.
This is the first time black is used in the project, so this PR also reformats the code with black.

The noticeable files are:
 - .flake8: a flake8 configuration. 
    Used by pre-commit but convenient for users too.
    Added ignore E203, which normally prohibits whitespace around the colon.
    Black inserts white-space around the colon sometimes as per PEP ([Whitespace in expressions and statements](https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements)).
    I could not find where `examples` where before explicitly ignored for flake8, but the develop branch examples raised problems with flake8, so I assumed we are excluding this folder.

 - mypy.ini: a mypy configuration. Used by pre-commit but convenient for users too.

 - .pre-commit-config.yaml: the pre-commit configuration
    I fixed the versions so everyone works with the same checkers. These are the versions used by GAMA, but not chosen for any particular reason (they were most recent at the time, I believe). 

 - contribution guidelines in docs

### CODE REVIEW

**Comment by PGijsbers** [comment]
I am getting some error from building docs/not finding a trace but I think they should be unrelated to this PR? Other than that I think it's good.

**Comment by mfeurer** [comment]
>I am getting some error from building docs/not finding a trace but I think they should be unrelated to this PR? Other than that I think it's good.

This was fixed by @Neeratyoy in #923

File: examples/40_paper/2015_neurips_feurer_example.py [hunk 1]
```diff
@@ -24,6 +24,7 @@
 ####################################################################################################
 # List of dataset IDs given in the supplementary material of Feurer et al.:
 # https://papers.nips.cc/paper/5872-efficient-and-robust-automated-machine-learning-supplemental.zip
+# fmt: off
```
Comment (mfeurer) [inline examples/40_paper/2015_neurips_feurer_example.py:27]: Is this for black? What would happen if you turn it off?

File: examples/40_paper/2018_neurips_perrone_example.py [hunk 2]
```diff
@@ -65,17 +63,18 @@ def fetch_evaluations(run_full=False,
     eval_df : dataframe
     task_ids : list
     flow_id : int
-    '''
+    """
     # Collecting task IDs as used by the experiments from the paper
-    if flow_type == 'svm' and run_full:
+    # fmt: off
```
Comment (mfeurer) [inline examples/40_paper/2018_neurips_perrone_example.py:68]: Same as above I guess.

File: openml/runs/functions.py [hunk 3]
```diff
@@ -439,15 +450,14 @@ def _run_task_get_arffcontent(
...
+            sample_no,
         )
 
-        (
-            pred_y,
-            proba_y,
-            user_defined_measures_fold,
-            trace,
-        ) = extension._run_model_on_fold(
+        (pred_y, proba_y, user_defined_measures_fold, trace,) = extension._run_model_on_fold(
```
Comment (mfeurer) [inline openml/runs/functions.py:460]: The brackets here can go away, right?

File: examples/40_paper/2015_neurips_feurer_example.py (hunk 1 above)
Comment (PGijsbers) [reply examples/40_paper/2015_neurips_feurer_example.py:27]: Yes it is for Black. It would put each element of the list in a separate row.

**Comment by PGijsbers** [comment]
I'll update the PR tomorrow. If after that only small stuff remains, would you be able to handle the rest of the PR? I have vacation this week :/ (not great timing, but I figured the pre-commit will help us with the hackathon).

**Comment by mfeurer** [comment]
> I'll update the PR tomorrow. If after that only small stuff remains, would you be able to handle the rest of the PR?

Thanks a lot and sure, I can do that.

> I have vacation this week

Enjoy :)

**Comment by PGijsbers** [comment]
I think this should have done the trick 👍  (rebase and processed feedback)

**Pull request merged**